databricks/koalas/series.py

```diff
@@ -4772,20 +4772,49 @@ def dot(self, other) -> Union[Scalar, "Series"]:
 
         >>> s @ s
         14
+
+        >>> kdf = ks.DataFrame({'x': [0, 1, 2, 3], 'y': [0, -1, -2, -3]})
+        >>> kdf
+           x  y
+        0  0  0
+        1  1 -1
+        2  2 -2
+        3  3 -3
+
+        >>> with ks.option_context("compute.ops_on_diff_frames", True):
+        ...     s.dot(kdf)
+        ...
+        x    14
+        y   -14
+        dtype: int64
         """
         if isinstance(other, DataFrame):
-            raise ValueError(
-                "Series.dot() is currently not supported with DataFrame since "
-                "it will cause expansive calculation as many as the number "
-                "of columns of DataFrame"
-            )
-        if self._kdf is not other._kdf:
-            if len(self.index) != len(other.index):
+            if not same_anchor(self, other) and not self.index.sort_values().equals(
+                other.index.sort_values()
+            ):
                 raise ValueError("matrices are not aligned")
-        if isinstance(other, Series):
-            result = (self * other).sum()
 
-        return result
+            other = other.copy()
+            column_labels = other._internal.column_labels
+
+            self_column_label = verify_temp_column_name(other, "__self_column__")
+            other[self_column_label] = self
```

Comment:
This raises `ValueError` when the `compute.ops_on_diff_frames` option is set to `False`.

I'm not sure we should use `option_context` here like `combine_first`.

https://github.com/databricks/koalas/blob/138c7b834b5f00d4d96c8ffb272bc1b1e025bd9c/databricks/koalas/series.py#L4710-L4711

Reply:
Oh, sorry we already check `same_anchor` above.

Please ignore this comment.

Reply:
I wonder if there are a lot of places like this. If there are a lot of such instances to fix, maybe we should just fix all to fail with an exception and let users switch on the configuration.

Reply:
I modified this PR to check option_context in the beginning (as 10127ac).

I did a rough grep and found two such places
```
$ git grep "with option_context(\"compute.ops_on_diff_frames\", True):"
databricks/koalas/series.py:            with option_context("compute.ops_on_diff_frames", True):
databricks/koalas/series.py:        with option_context("compute.ops_on_diff_frames", True):
```

Shall we change them to raise an exception instead?

Reply:
Also
```
databricks/koalas/namespace.py:            with ks.option_context("compute.ops_on_diff_frames", True):
```

Reply:
I'm fine with disabling implicit switch-on (in the separate PR).

Reply:
Sounds good! I will be working on this.